### PR TITLE
Task page : task filters

### LINKED ISSUE #9582: sorting Dailies in "All" filter causes "Due" filter to be applied
As reported in the Report a Bug guild by Il Mago Azzurro (c073342f-4a65-4a13-9ffd-9e7fa5410d6b):

"Website, trying to reorder my dailies with the "All" filter active results in having the "Due" filter (the default one) activating everytime the page sync after I move something, I had to move only 2 things, but it was really longer than it should be..."

I've confirmed that in my account, although it didn't happen every time (just most of the time).

The correct behaviour is that when you are sorting Dailies using any of the All/Due/NotDue filters, the filter should never be changed automatically.

### PR DESCRIPTION
**Pull request by kartikadur**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes https://github.com/HabitRPG/habitica/issues/9582

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

All changes made to reduce response / processing times, improve data reactivity, and reduce data state ambiguity. I have run tests and checked things multiple time, but I don't expect them to be exhaustive. 

#### Existing files:
TaskColumn (column.vue):
+ store.getters access where possible instead of store.state
+ can only access its own task type instead of all tasks
+ task Labels and Filters assigned to own file (client/libs/store/helpers/filterTasks.js)
  - task label, filter labels, and active filter function assigned when component is created
+ functions adjusted where needed to incorporate above changes
+ unit tests for functions where possible

Task-Getters (store/getters/tasks.js)
+ returns all tasks of requested type
+ returns all tasks of requested type filtered by categories (all, due, notDue for dailies, etc.)
+ unit tests

User-Getters (store/getters/user.js)
+ return user buffs, preferences, and task order by task type

#### New files:
Filter-Tasks (client/libs/store/helpers/filterTasks.js)
+ contains all task filter data and return task label, filter labels, and active filter
+ unit tests

Order-Tasks (client/libs/store/helpers/orderTasks.js)
+ assigns task order by single task type for efficiency
+ unit tests

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID:

### CODE REVIEW

**Comment by TheHollidayInn** [comment]
This is looking pretty great! Thank you!

Before I review this, am I correct in seeing that you have removed the ability to set a filter based on the user's preference for now? I see your todos.

**Comment by kartikadur** [comment]
The `taskFilters` object in filterTasks.js and `getFilteredTaskList()` function in tasks.js, in the code below, show how a user's preferences are passed to the filters. Is this what you are referencing?

```javascript
daily: {
  label: 'dailies',
  filters: [
    { label: 'all', filterFn: () => true, default: true },
    { label: 'due', filterFn: userPrefs => t => !t.completed && shouldDo(new Date(), t, userPrefs) },
    { label: 'notDue', filterFn: userPrefs => t => t.completed || !shouldDo(new Date(), t, userPrefs) },
  ],
},
```

```javascript
if (type === 'daily' && (filterType === 'due' || filterType === 'notDue')) {
  selectedFilter = {
    ...selectedFilter,
    filterFn: selectedFilter.filterFn(userPreferences),
  };
}
```

As far as I know only `dailies - due and notDue` required them and not `todos`.

Let me know if there were any changes to the filter functions for `todos` and I can update them accordingly before the review.

**Comment by TheHollidayInn** [comment]
Oh yep! That looks good to me :D

**Comment by kartikadur** [comment]
oh good, I thought I had messed up and forgotten something. :)

**Comment by paglias** [comment]
Going to test this now @kartikadur . Thanks a lot for the tests, although we have them currently disabled for the client, pending a fix for https://github.com/HabitRPG/habitica/issues/9276

File: website/client/components/tasks/column.vue [hunk 1]
```diff
@@ -591,35 +567,36 @@ export default {
...
+      // filter requested tasks by search text
+      if (!isEmpty(searchText)) {
+        // to ensure broadest case insensitive search matching
+        let searchTextLowerCase = searchText.toLowerCase();
+        filteredTaskList = taskList.filter(
+          task => {
+            // eslint rule disabled for block to allow nested binary expression
+            /* eslint-disable no-extra-parens */
+            return (
+              (!isEmpty(task.text) && task.text.toLowerCase().indexOf(searchTextLowerCase) > -1) ||
```
Comment (paglias) [inline website/client/components/tasks/column.vue]: why ar you using `!isEmpty` instead of `task.note && ...` these are simple strings, not objects

Comment (paglias) [reply website/client/components/tasks/column.vue]: And the check for `task.text` isn't necessary since we enforce its existence

File: website/client/libs/store/helpers/filterTasks.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      } else {
+        return filterFunction;
+      }
+    }
+  };
+}
+
+const getActiveFilter = activeFilter(taskFilters);
+
+export {
```
Comment (paglias) [inline website/client/libs/store/helpers/filterTasks.js]: these should be exported separately not as a single object

File: website/client/libs/store/helpers/filterTasks.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+const getFilterLabels = filterLabel(taskFilters);
+
+function activeFilter (filterList) {
+  return (type, filterType = '') => {
+    let filterListByType = filterList[type].filters;
+    if (isEmpty(filterType)) {
+      return filterListByType.find(f => f.default === true);
+    } else {
+      // check if filter type is available, else send default filter
```
Comment (paglias) [inline website/client/libs/store/helpers/filterTasks.js]: is this necessary? I mean, we control which filters are activable it should never happen for the user to select a filter that's unavaible

File: website/client/libs/store/helpers/orderTasks.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import compact from 'lodash/compact';
+import isEmpty from 'lodash/isEmpty';
+
+// sets task order for single task type only.
+// Accepts task list and corresponding taskorder for its task type.
+export function orderSingleTypeTasks (rawTasks, taskOrder = []) {
+  // if there is no predefined task order return task list as is.
+  if (isEmpty(taskOrder)) return rawTasks;
+  // // eslint-disable-next-line no-console
```
Comment (paglias) [inline website/client/libs/store/helpers/orderTasks.js]: let's remove thse two lines

File: website/client/libs/store/helpers/orderTasks.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import compact from 'lodash/compact';
+import isEmpty from 'lodash/isEmpty';
+
+// sets task order for single task type only.
+// Accepts task list and corresponding taskorder for its task type.
+export function orderSingleTypeTasks (rawTasks, taskOrder = []) {
+  // if there is no predefined task order return task list as is.
+  if (isEmpty(taskOrder)) return rawTasks;
```
Comment (paglias) [inline website/client/libs/store/helpers/orderTasks.js]: at this point it's not necessary to pass a default to taskOrder, let's remove that and simply do `if (!taskOrder)`

File: website/client/store/getters/tasks.js [hunk 6]
```diff
@@ -109,3 +116,51 @@ export function getTaskClasses (store) {
...
+  return ({
+    type,
+    filterType = '',
+  }) => {
+    // get requested tasks
+    // check if task list has been passed as override props
+    // assumption: type will always be passed as param
+    let requestedTasks = getters['tasks:getUnfilteredTaskList'](type);
+
+    let userPreferences = state.user.data.preferences || {};
```
Comment (paglias) [inline website/client/store/getters/tasks.js]: these defaults aren't necessary, the api always return these fields

File: website/client/store/getters/tasks.js [hunk 7]
```diff
@@ -109,3 +116,51 @@ export function getTaskClasses (store) {
...
+        filterFn: selectedFilter.filterFn(userPreferences),
+      };
+    }
+
+    requestedTasks = requestedTasks.filter(selectedFilter.filterFn);
+    if (selectedFilter.sort) {
+      requestedTasks = sortBy(requestedTasks, selectedFilter.sort);
+    }
+
+    // // eslint-disable-next-line no-console
```
Comment (paglias) [inline website/client/store/getters/tasks.js]: let's remove these lines

File: website/client/components/tasks/column.vue [hunk 8]
```diff
@@ -339,45 +310,49 @@ export default {
...
     }),
-    onUserPage () {
-      let onUserPage = Boolean(this.taskList.length) && (!this.taskListOverride || this.taskListOverride.length === 0);
+    ...mapGetters({
+      getFilteredTaskList: 'tasks:getFilteredTaskList',
+      getUnfilteredTaskList: 'tasks:getUnfilteredTaskList',
+      getUserPreferences: 'user:preferences',
+      getUserBuffs: 'user:buffs',
+    }),
+    // Temporarily commented till definitely known if onUserPage() is needed or not
```
Comment (paglias) [inline website/client/components/tasks/column.vue]: Looks like it's never used (or was before this PR) so we can remove this

**Review by paglias — CHANGES_REQUESTED** [review]
Left some comments, in general it seems to work well. Let's just not use `isEmpty` when dealing with strings :)

File: website/client/components/tasks/column.vue (hunk 8 above)
Comment (kartikadur) [reply website/client/components/tasks/column.vue]: I want to remove this bit of code, but wasn't sure if it was being used anywhere else. If it isn't I'll take it out.

File: website/client/components/tasks/column.vue (hunk 1 above)
Comment (kartikadur) [reply website/client/components/tasks/column.vue]: Force of habit trying to be thorough.

Comment (kartikadur) [reply website/client/components/tasks/column.vue]: I tend to go overboard sometimes, thanks for the heads up

File: website/client/libs/store/helpers/filterTasks.js (hunk 3 above)
Comment (kartikadur) [reply website/client/libs/store/helpers/filterTasks.js]: activeFilter (and by extention getActiveFilter) takes two params. param 1 is the task type, while the second is optional filter type (e.g. all, due, notDue for dailies, etc.). I'll simplify the structure and remove the unit test checks for boundary conditions for unavailable filters.

File: website/client/libs/store/helpers/filterTasks.js (hunk 2 above)
Comment (kartikadur) [reply website/client/libs/store/helpers/filterTasks.js]: updated as separate exports

File: website/client/libs/store/helpers/orderTasks.js (hunk 5 above)
Comment (kartikadur) [reply website/client/libs/store/helpers/orderTasks.js]: updated as suggested

File: website/client/store/getters/tasks.js (hunk 7 above)
Comment (kartikadur) [reply website/client/store/getters/tasks.js]: updated as suggested

File: website/client/store/getters/tasks.js (hunk 6 above)
Comment (kartikadur) [reply website/client/store/getters/tasks.js]: updated as suggested

File: website/client/libs/store/helpers/orderTasks.js (hunk 4 above)
Comment (kartikadur) [reply website/client/libs/store/helpers/orderTasks.js]: updated as suggested

**Comment by kartikadur** [comment]
@paglias thanks for the updates. I have a bad habit of overthinking boundary cases / worst case scenarios. So its really helpful getting pointers like these.

File: website/client/components/tasks/column.vue (hunk 8 above)
Comment (paglias) [reply website/client/components/tasks/column.vue]: well, it's a computed property, isn't it? I just searched if it was used anywhere in the component but let me know if I'm missing something

Comment (kartikadur) [reply website/client/components/tasks/column.vue]: It is neither used in the same component nor in the direct parent.

onUserPage() resets the dailies filter to default and rewards filter to empty array if the user is not on their own page. Not sure if this type of functionality was designed.

Comment (paglias) [reply website/client/components/tasks/column.vue]: cc @TheHollidayInn

**Comment by paglias** [comment]
@kartikadur thanks for this, as for the other PRs if you have an Habitica UUID let me know

**Comment by paglias** [comment]
Hi @kartikadur we've noticed that the user's own tasks are now appearing in the group plan tasks page. I think this is due to this PR, can you give a look? Thanks!

**Comment by kartikadur** [comment]
@paglias I'm looking into it. How urgent is this required change?

**Comment by paglias** [comment]
We have to find a solution by tomorrow, I can look into it this evening or
will revert the pr in the meantime

Il 21 Feb 2018 16:08, "kartik adur" <notifications@github.com> ha scritto:

**Comment by kartikadur** [comment]
@paglias is there a way to do group plan tasks in a dev instance? I had accounted for the tasklistoverride in my code but I may need it to check how the code change works to be certain of where the problem is.

> the user's own tasks are now appearing in the group plan tasks page

are the user's tasks showing up instead of group-plan tasks or are they combined with the group-plan tasks?

**Comment by paglias** [comment]
They're combined. You should be able to create a free group plan by trying
to create a new one and when you go to pay Stripe should have a yello test
badge somewhere, then you can use 4242 4242 4242 4242 with and CVC and
expiration date and it should work

Il 21 Feb 2018 16:18, "kartik adur" <notifications@github.com> ha scritto:

**Comment by kartikadur** [comment]
Thanks.

For the moment reverting the PR till I have a chance to test and update the code would be prudent

**Comment by kartikadur** [comment]
A preliminary check indicates that there should not be a problem of tasks merging in the frontend since group tasks and user tasks are treated as mutually exclusive. Any group tasks passed as props overwrite the user's tasks.

Could you provide any conditions you used for running this test / scenario where you ended up with merged group and user tasks? is there a possibility that the group + user tasks are retreived in a merged state from the API?

**Comment by paglias** [comment]
Whenever you open a group page I see group tasks + user tasks (dailys always, habits appear briefly and then disappear) I checked and the api is not returning the user tasks with the group ones

schermata_2018-02-21_alle_17_34_37
schermata_2018-02-21_alle_17_34_57

**Comment by paglias** [comment]
It looks like that:

1. If you have no group dailies, the user dailies are rendered
2. If you have group habits (for example) the user habits are renderend for just 1 sec and then replaced by the group habits

**Comment by kartikadur** [comment]
Thanks for the update. I think I may have figured out where the problem might be. It'll take me a day or two to update the code and submit changes.

Should I submit the changes to branch that created this PR or should I make a new PR?

**Comment by paglias** [comment]
New PR, i can also look into it if you don't have time

**Comment by paglias** [comment]
Could the problem be in this line https://github.com/HabitRPG/habitica/pull/9898/files#diff-55116909dfc8ad199c0b7a35ba675675R334 ? I feel like it should have a condition in case we're on a group

**Comment by kartikadur** [comment]
I'm attaching the changes I'll be making in the next 48 hours. 
change 1: Line 283
```JavaScript
props: ['type', 'isUser', 'searchText', 'selectedTags', 'taskListOverride', 'group'], // @TODO: maybe we should store the group on state?
```
set 'isUser' prop to have a default value. I would suggest setting it to default `false` as it would require fewer changes in the overall code.


change 2: Line 334 - 339
```JavaScript
let filteredTaskList = isEmpty(this.taskListOverride) ?
  this.getFilteredTaskList({
    type: this.type,
    filterType: this.activeFilter.label,
  }) :
  this.taskListOverride;
```
in these lines check for 'this.isUser' as `true` or `false` instead of 'this.taskListOverride' as empty or not.

**Comment by kartikadur** [comment]
Let me know if its faster for you to make the changes, in which case I will leave it to you.

**Comment by Alys** [comment]
@kartikadur We'll revert the PR as you suggest so that this doesn't have to be rushed. Thank you for looking into it! Please keep going if you'd like to!

**Comment by kartikadur** [comment]
Ok, I'll have the new PR up by Friday.

**Comment by paglias** [comment]
Reverted the PR in https://github.com/HabitRPG/habitica/commit/d5e4be85e991740e6ec622c41f7c38a307b802ea
